include/mbedtls/rsa.h

```diff
@@ -894,6 +894,15 @@ int mbedtls_rsa_rsassa_pkcs1_v15_sign( mbedtls_rsa_context *ctx,
  *                 Specifications</em> it is advised to keep both hashes the
  *                 same.
  *
+ * \note           This function calculates and uses the largest possible salt
+ *                 size. Normally this is the hash size, which is the maximum
+ *                 size the salt can have. If there is not enough room, use the
+ *                 maximum salt size that fits, but always use at least hash
+ *                 size - 2 bytes of salt. The constraint is that the hash size
+ *                 plus the salt size plus 2 bytes must be at most the key
+ *                 size. This complies with FIPS 186-4 §5.5 (e) and RFC 8017
+ *                 (PKCS#1 v2.2) §9.1.1 step 3.
```

Comment:
This is not completely accurate and a little confusing regarding “the constraint”, sorry. The salt can be longer than the hash, it's allowed by PKCS#1 but forbidden by FIPS. Suggestion:

> This function always uses the maximum salt size `slen` that satisfies the following constraints:

> * `slen + hlen + 2 <= siglen` where `hlen` is the length of the hash (`\p hashlen` if `\p md_alg` is `#MBEDTLS_MD_NONE`, otherwise the length of `md_alg`) and `siglen` is the length of the signature (which is also the length of the key).
> * `slen <= hlen`.
> * `slen >= hlen - 2`.
>
> The first two constraints ensure that the salt length complies with FIPS 186-4 §5.5 (e) and RFC 8017 (PKCS#1 v2.2) §9.1.1 step 3.

Or:

> This function always uses the maximum possible salt size, up to the length of the payload hash. This choice of salt size complies with FIPS 186-4 §5.5 (e) and RFC 8017 (PKCS#1 v2.2) §9.1.1 step 3. Furthermore this function enforces a minimum salt size which is the hash size minus 2 bytes. If this minimum size is too large given the key size (the salt size, plus the hash size, plus 2 bytes must be no more than the key size in bytes), this function returns `#MBEDTLS_ERR_RSA_BAD_INPUT_DATA`.

Reply:
I've gone with the second suggestion. Thanks.